Adding a decorator to assert a user is not logged

Micha546 opened the pull request:
# Description
Adding a decorator to assert a user is not logged

Added the decorator in main.decorators.
Added the new decorator to the register views of Seeker
and Apartment.
Added tests to assert the decorator is working properly.

This PR is dependent on #147 - merged

used following refernces:
https://www.youtube.com/watch?v=eBsc65jTKvw&t=766s
https://docs.djangoproject.com/en/3.2/ref/contrib/auth/
https://www.datacamp.com/community/tutorials/decorators-python

## Screenshots
![image](https://user-images.githubusercontent.com/79044795/117153431-a1389500-adc3-11eb-84fe-187d3267d621.png)


## Manual Tests
Did some tests to see that the decorator is working as intended

## Additional Information
Found a bug/issue with flake8, when i tried to import a fixture and use it flake8 was confused:
![flake_error](https://user-images.githubusercontent.com/79044795/117153664-de9d2280-adc3-11eb-923b-bca674b684c1.png)

We will probably have to redisign the fixtures in all the tests files, for now I added  `# noqa: F811` and `# noqa: F401` to
make flake8 ignore those false positives...
https://stackoverflow.com/questions/47196947/py-test-deal-with-both-pylint-and-flake8-when-importing-features-from-a-module

### Issues close by this PR
Fix #137

Omeramsc commented on seekers/tests.py:
```diff
@@ -1,6 +1,7 @@
 from apartments.models import City
 from .forms import SeekerCreationForm
 from django.urls import reverse
+from users.tests import user_model  # noqa: F401
```
You should not import fixtures but put them inside a conftest file. with that, they'll be automatically available for use across the scope they're in.

read more:
https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session

Don't ignore the linting issues, they're there for a reason.

Micha546 replied:
Yes, I saw that. but I think that moving all the fixtures to a new file and changing all the tests in the process is not in the scope of this PR. Can I open an issue for the redesigning of the fixtures and the tests and have this PR merged?

Micha546 replied:
As I mentioned in the Additional Information part:
Found a bug/issue with flake8, when i tried to import a fixture and use it flake8 was confused,
We will probably have to redisign the fixtures in all the tests files, for now I added # noqa: F811 and # noqa: F401 to
make flake8 ignore those false positives...

Omeramsc replied:
No, what you should do in that case is first to open a PR that fixes this issue, and then rebase this PR on top of it.

Micha546 replied:
Ok, I'll start working on it
👍

Micha546 replied:
@Omeramsc I pushed a new PR that moves everything to conftest.py

Micha546 committed:
Adding a decorator to assert a user is not logged

Added the decorator in main.decorators.
Added the new decorator to the register views of Seeker
and Apartment.
Added tests to assert the decorator is working properly.

used following refernces:
https://www.youtube.com/watch?v=eBsc65jTKvw&t=766s
https://docs.djangoproject.com/en/3.2/ref/contrib/auth/

simzacks commented on apartments/tests.py:
```diff
@@ -1,85 +1,47 @@
-        assert new_city.cityName == 'some_city'
+    def test_city_creation(self, city_model):
+        assert city_model is not None
+        assert city_model.cityName == 'nice_city'
```
this implicitly verifies that city_model is not None, so no need to explicitly check .
If city_model is None (or any object that doesn't have a city_name attribute) you'll get an AttributeError exception
`assert city_model is not None`
As an aside, if you're testing for not None you can leave that part out `assert city_model`

Micha546 replied:
@simzacks Those changes are part of the #147 PR which this PR is dependent on.
I didn't write any new tests just moved fixtures to conftest.py

nevertheless I'll look at the changes you requested and change the code accordingly in #147

simzacks commented on apartments/tests.py:
```diff
@@ -1,85 +1,47 @@
-        assert check_city == new_city
+    def test_get_city(self, city_model):
+        check_city = City.objects.get(cityName=city_model.cityName)
+        assert check_city == city_model
```
I'm unclear what you're checking here.
If you're trying to see if Django's get function works. There's no need to test that.

Micha546 replied:
Ok, ill delete this test in #147

simzacks commented on apartments/tests.py:
```diff
@@ -1,85 +1,47 @@
+        assert apartment_model is not None
+        assert apartment_model.address == 'street'
+        assert apartment_model.rent == 4500
+        assert apartment_model.is_relevant is True
```
as above, no need to assert apartment_model is not none as it is implicit.
no need to use `is True`, `assert apartment_model.is_relevant` will do the same thing

simzacks commented on apartments/tests.py:
```diff
@@ -1,85 +1,47 @@
-        assert check_apartment == new_apartment
+    def test_get_apartment(self, apartment_model):
+        check_apartment = Apartment.objects.get(owner=apartment_model.owner)
+        assert check_apartment == apartment_model
```
same as above, what are you testing here?

Micha546 replied:
I'll delete this test in #147

simzacks commented on apartments/tests.py:
```diff
@@ -1,85 +1,47 @@
-    def test_get_apartment_by_id(self, new_apartment):
-        check_apartment_success = Apartment.get_apartment_by_id(new_apartment.owner.id)
+    def test_get_apartment_by_id(self, apartment_model):
+        check_apartment_success = Apartment.get_apartment_by_id(apartment_model.owner.id)
```
Here you are testing a function that you wrote `get_apartment_by_id` so that is a good test.

Micha546 replied:
Not my test, I only changed the fixture he is getting 👍

simzacks commented on apartments/tests.py:
```diff
@@ -1,85 +1,47 @@
-    def test_update_apartment_view_to_owner(self, client, new_apartment):
-        client.login(email='test@test.com', password='password')
+    def test_update_apartment_view_to_owner(self, client, apartment_model):
+        client.login(email='apartmentemail@address.com', password='password')
```
You are passing in apartment_model as a parameter and it isn't used.

Micha546 replied:
I think that without it the client wouldn't be able to login, but I'm not sure, it's not my test but ill try to run it without the apartment_model

Micha546 replied:
![image](https://user-images.githubusercontent.com/79044795/117565407-b3fbe400-b0b9-11eb-9e04-5b2f93426db7.png)

I've tried running it without the fixture and it failed.

simzacks commented on apartments/tests.py:
```diff
@@ -1,85 +1,47 @@
-    def test_update_apartment_view_to_not_owner_user(self, client, new_user):
-        client.login(email='test@test.com', password='password')
+    def test_update_apartment_view_to_not_owner_user(self, client, user_model):
+        client.login(email='email@address.com', password='password')
```
You are passing in user_model as a parameter and it isn't used.

Micha546 replied:
same as above, ill try running it without it

Micha546 replied:
Does not work without the fixture

simzacks commented on apartments/tests.py:
```diff
@@ -118,32 +80,29 @@ def test_apartment_form_validity(city, address, rent, num_of_roomates, num_of_ro
+def test_fail_to_save_apartment_form_with_commit_true(valid_apartment_creation_form):
     with pytest.raises(ValueError):
-        valid_form.save(commit=True)
+        valid_apartment_creation_form.save(commit=True)
```
Is the `valid_apartment_creation_form` not valid? Why is it supposed to raise an exception?

Micha546 replied:
With commit=True it's supposed to save the Apartment immediately to the data-base, because the form does not have a `owner` field I made it so the form throws an exception when you try to save it with commit=True, this test tests it

simzacks commented on apartments/tests.py:
```diff
@@ -118,32 +80,29 @@ def test_apartment_form_validity(city, address, rent, num_of_roomates, num_of_ro
+
+
+@pytest.mark.django_db
+def test_accessing_apartment_register_view_with_logged_user(client, user_model):
```
why is user_model being passed in?

Micha546 replied:
For the login, I'll try running it without it

Micha546 replied:
Does not work without the fixture

simzacks commented on main/decorators.py:
```diff
@@ -0,0 +1,12 @@
+from django.shortcuts import redirect
```
Does this built in django decorator do what you're trying to do here?
https://docs.djangoproject.com/en/3.2/topics/auth/default/#the-login-required-decorator

Micha546 replied:
I wrote the opposite decorator, if you're logged in, you can't access the page

simzacks commented on main/tests.py:
```diff
@@ -37,8 +21,8 @@ def test_enter_logout_page(self, client):
-    def test_successful_login_and_logout(self, client, make_user):
-        sample_user = make_user("test1@gmail.com", "testy", "test", "1995-05-05", "test123")
+    def test_successful_login_and_logout(self, client):
+        sample_user = get_user_model().objects.create_user("test1@gmail.com", "testy", "test", "1995-05-05", "test123")
```
You're creating the same user in all the tests here. Won't that cause a problem?
You should probably use a setup function and create the user once

Micha546 replied:
`make_user` was a fixture that did exactly what `get_user_model().objects.create_user` is doing, so I changed it

simzacks commented on seekers/tests.py:
```diff
@@ -41,32 +33,29 @@ def test_seeker_form_validity(city, start_date, min_rent, max_rent, num_of_rooma
 
 @pytest.mark.django_db
-def test_fail_to_save_seeker_form_with_commit_true(valid_form):
+def test_fail_to_save_seeker_form_with_commit_true(valid_seeker_creation_form):
```
why are you expecting this to fail?

Micha546 replied:
Yes, this form too has no `base-user` field so I made it throw an exception when trying to save with commit=True

simzacks commented on seekers/tests.py:
```diff
@@ -41,32 +33,29 @@ def test_seeker_form_validity(city, start_date, min_rent, max_rent, num_of_rooma
+
+
+@pytest.mark.django_db
+def test_accessing_seeker_register_view_with_logged_user(client, user_model):
```
why pass in the user_model?

Micha546 replied:
For log-in ill try running it without it

Micha546 replied:
Does not work without the fixture

Micha546 commented:
Just another heads-up, this PR is dependent on #147, so please merge #147 first

simzacks commented:
Looks good. I'll wait until 147 is merged and this is rebased before approving
